homeassistant/components/volvooncall/__init__.py

```diff
@@ -1,145 +1,80 @@
-                        cv.time_period, vol.Clamp(min=DEFAULT_UPDATE_INTERVAL)
-                    ),  # ignored, using DataUpdateCoordinator instead
-                    vol.Optional(CONF_NAME, default={}): cv.schema_with_slug_keys(
-                        cv.string
-                    ),  # ignored, users can modify names of entities in the UI
-                    vol.Optional(CONF_RESOURCES): vol.All(
-                        cv.ensure_list, [vol.In(RESOURCES)]
-                    ),  # ignored, users can disable entities in the UI
-                    vol.Optional(CONF_REGION): cv.string,
-                    vol.Optional(CONF_SERVICE_URL): cv.string,
-                    vol.Optional(CONF_MUTABLE, default=True): cv.boolean,
-                    vol.Optional(CONF_SCANDINAVIAN_MILES, default=False): cv.boolean,
-                }
-            ),
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Migrate from YAML to ConfigEntry."""
+    if DOMAIN not in config:
+        return True
+
+    hass.data[DOMAIN] = {}
+
+    if not hass.config_entries.async_entries(DOMAIN):
+        new_conf = {}
+        new_conf[CONF_USERNAME] = config[DOMAIN].get(CONF_USERNAME)
+        new_conf[CONF_PASSWORD] = config[DOMAIN].get(CONF_PASSWORD)
+        new_conf[CONF_REGION] = config[DOMAIN].get(CONF_REGION)
+        new_conf[CONF_SCANDINAVIAN_MILES] = config[DOMAIN].get(CONF_SCANDINAVIAN_MILES)
+        new_conf[CONF_MUTABLE] = config[DOMAIN].get(CONF_MUTABLE)
+        if new_conf[CONF_MUTABLE] is None:
+            new_conf[CONF_MUTABLE] = True
+
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=new_conf
+            )
         )
-    },
-    extra=vol.ALLOW_EXTRA,
-)
 
+    return True
 
-async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Set up the Volvo On Call component."""
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: config_entries.ConfigEntry
+) -> bool:
+    """Set up the Volvo On Call component from a ConfigEntry."""
     session = async_get_clientsession(hass)
 
     connection = Connection(
         session=session,
-        username=config[DOMAIN].get(CONF_USERNAME),
-        password=config[DOMAIN].get(CONF_PASSWORD),
-        service_url=config[DOMAIN].get(CONF_SERVICE_URL),
-        region=config[DOMAIN].get(CONF_REGION),
+        username=entry.data[CONF_USERNAME],
+        password=entry.data[CONF_PASSWORD],
+        service_url=None,
+        region=entry.data[CONF_REGION],
     )
 
-    hass.data[DATA_KEY] = {}
+    hass.data.setdefault(DOMAIN, {})
 
-    volvo_data = VolvoData(hass, connection, config)
+    volvo_data = VolvoData(hass, connection, entry)
 
-    hass.data[DATA_KEY] = VolvoUpdateCoordinator(hass, volvo_data)
+    hass.data[DOMAIN][entry.entry_id] = VolvoUpdateCoordinator(hass, volvo_data)
 
     return await volvo_data.update()
```

Comment:
We want to call `coordinator.async_config_entry_first_refresh` instead.

The coordinator update method should raise `ConfigEntryAuthFailed` if authentication fails and otherwise raise `UpdateFailed`.